x/ccv/provider/keeper/key_assignment.go

```diff
@@ -0,0 +1,618 @@
+		providerAddr := utils.TMCryptoPublicKeyToConsAddr(update.PubKey)
+		if consumerKey, found := k.GetValidatorConsumerPubKey(ctx, chainID, providerAddr); found {
+			newUpdates[i].PubKey = consumerKey
+		}
+	}
+	return newUpdates
+}
+
+// ApplyKeyAssignmentToValUpdates applies the key assignment to the validator updates
+// received from the staking module
+func (k Keeper) ApplyKeyAssignmentToValUpdates(
+	ctx sdk.Context,
+	chainID string,
+	valUpdates []abci.ValidatorUpdate,
+) (newUpdates []abci.ValidatorUpdate, err error) {
+	for _, valUpdate := range valUpdates {
+		providerAddr := utils.TMCryptoPublicKeyToConsAddr(valUpdate.PubKey)
+
+		// If a key assignment replacement is found, we remove the valupdate with the old consumer key,
+		// create tow new valupdates,
+		//  - setting the old consumer key's power to 0
+		//  - and setting the new consumer key's power to the power in the update
+		oldAssignment, found := k.GetKeyAssignmentReplacement(ctx, chainID, providerAddr)
+		if found {
+			newUpdates = append(newUpdates, abci.ValidatorUpdate{
+				PubKey: oldAssignment.PubKey,
+				Power:  0,
+			})
+
+			newConsumerKey, found := k.GetValidatorConsumerPubKey(ctx, chainID, providerAddr)
+			if !found {
+				return newUpdates, sdkerrors.Wrapf(types.ErrConsumerKeyNotFound, "consumer key not found for %s", providerAddr)
+			}
+			newUpdates = append(newUpdates, abci.ValidatorUpdate{
+				PubKey: newConsumerKey,
+				Power:  valUpdate.Power,
+			})
+			k.DeleteKeyAssignmentReplacement(ctx, chainID, providerAddr)
+		} else {
+			// If a pending key assignment is not found, we check if the validator's key is assigned.
+			// If it is, we replace the update containing the provider key with the update containing the consumer key.
+			consumerKey, found := k.GetValidatorConsumerPubKey(ctx, chainID, providerAddr)
+			if found {
+				newUpdates = append(newUpdates, abci.ValidatorUpdate{
+					PubKey: consumerKey,
+					Power:  valUpdate.Power,
+				})
+			} else {
+				// keep the same update
+				newUpdates = append(newUpdates, valUpdate)
+			}
+		}
+	}
+
+	// For any key assignment replacements that did not have a corresponding validator update already,
+	// set the old consumer key's power to 0 and the new consumer key's power to the
+	// power in the pending key assignment.
+	var addrToRemove []sdk.ConsAddress
+	k.IterateKeyAssignmentReplacements(ctx, chainID, func(
+		pAddr sdk.ConsAddress,
+		pendingKeyAssignment abci.ValidatorUpdate,
+	) (stop bool) {
+		addrToRemove = append(addrToRemove, pAddr)
+		newUpdates = append(newUpdates, abci.ValidatorUpdate{
+			PubKey: pendingKeyAssignment.PubKey,
+			Power:  0,
+		})
+
+		newConsumerKey, found := k.GetValidatorConsumerPubKey(ctx, chainID, pAddr)
+		if !found {
+			err = sdkerrors.Wrapf(types.ErrConsumerKeyNotFound, "consumer key not found for %s", pAddr)
+			return true
+		}
+		newUpdates = append(newUpdates, abci.ValidatorUpdate{
+			PubKey: newConsumerKey,
+			Power:  pendingKeyAssignment.Power,
```

Comment:
This is using the old name ` pendingKeyAssignment`

Reply:
Removed.